feat: multichain api e2e

abretonc7s opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR implements the infrastructure for multichain E2E testing focused specifically on EVM networks. The changes include:

1. Fixed static server URL handling for the multichain test dapp by adding path rewriting for `/test-dapp-multichain/` prefixes
2. Added proper configuration in fixture-helper.js to serve the multichain test dapp
3. Created initial E2E test example that loads the multichain test dapp

The goal is to replicate the core test modules from the MetaMask extension's multichain E2E tests, starting with basic dapp connectivity and gradually implementing the full suite of tests for wallet_createSession, wallet_invokeMethod, wallet_getSession, and wallet_revokeSession.

At this stage, we're focusing only on EVM networks as we're not yet ready for non-EVM network testing.


Run with:

```bash
# Load environment and run all multichain tests
source .e2e.env && MULTICHAIN=1 yarn test:e2e:ios:debug:run e2e/specs/multichain/wallet-createSession.spec.ts e2e/specs/multichain/wallet-getSession.spec.ts e2e/specs/multichain/wallet-invokeMethod.spec.ts e2e/specs/multichain/wallet-notify.spec.ts e2e/specs/multichain/wallet-revokeSession.spec.ts e2e/specs/multichain/wallet-sessionChanged.spec.ts

# Run individual test files
source .e2e.env && MULTICHAIN=1 yarn test:e2e:ios:debug:run e2e/specs/multichain/wallet-createSession.spec.ts
```


## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

ffmcgee725 committed:
refactor: reduce instead of filter and map

ffmcgee725 committed:
refactor: move functions from Engine to transaction utils controller

ffmcgee725 committed:
refactor: move to another util folder

ffmcgee725 committed:
format

ffmcgee725 committed:
tsdocs

ffmcgee725 committed:
refactor: updatePermittedChains rename

ffmcgee725 committed:
test: fix

ffmcgee725 committed:
Merge branch 'main' into feat/caip-25-refactor

ffmcgee725 committed:
test: e2e unskip problematic mock server issue test

ffmcgee725 committed:
minor fix

ffmcgee725 committed:
lint

ffmcgee725 committed:
lint

ffmcgee725 committed:
lint

ffmcgee725 committed:
fix background bridge

ffmcgee725 committed:
further fixes on background bridge test

ffmcgee725 committed:
dopcs

ffmcgee725 committed:
lint fix

jiexi committed:
cleanup chainsToPermit

ffmcgee725 committed:
lint

ffmcgee725 committed:
chore: set ownership of util/permissions to wallet-api-platform

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge branch 'feat/caip-25-refactor' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
Merge branch 'feat/caip-25-refactor' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
Merge branch 'main' into feat/caip-25-refactor

ffmcgee725 committed:
Merge branch 'feat/caip-25-refactor' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
refactor: move 'rejectOriginPendingApprovals' to permission utils file

ffmcgee725 committed:
test: fix app/core/RPCMethods/RPCMethodMiddleware.test.ts

ffmcgee725 committed:
Merge branch 'main' into feat/caip-25-refactor

Cal-L committed:
chore:  Report undefined selected account in migrations (#14823)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

Findings from undefined account ID continues to point to migrations
returning a state with an undefined selectedAccount property -
https://github.com/MetaMask/metamask-mobile/issues/14276. It's possible
that the property is also undefined prior to being migrated. In any
case, we're adding more logs to the migrations that mutate
selectedAccount to filter out the possibility of migrations producing an
undefined state. We've also fixed the error log where validation was not
reporting any issues regarding selectedAccount.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-mobile/issues/14822

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge remote-tracking branch 'origin/feat/caip-25-refactor' into feat/caip-25-refactor

Matt561 committed:
feat: STAKE-1006 add earn remote feature flags to mobile (#14660)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR adds remote feature flags for Earn experiences.

### Changes 
#### Added the following feature flags (and selectors)
- `earnPooledStakingEnabled`: Pooled-staking kill switch. When enabled,
hides all pooled-staking cta's and stake entrypoints. However, we do not
hide the "unstake" and "claim" buttons because we don't want to block
users from withdrawing funds (self-custody).
- `selectPooledStakingServiceInterruptionBannerEnabledFlag`: When
enabled, displays a "we're down for maintenance" banner for Ethereum and
Staked Ethereum assets.
- `selectStablecoinLendingEnabledFlag`: Lending kill switch. When
enabled, will hide all depositing entrypoints. We will be using this
soon when wiring up the lending experience.
- `selectStablecoinLendingServiceInterruptionBannerEnabledFlag`: When
enabled, displays a "we're down for maintenance" banner for supported
stablecoin lending assets. Not in use yet.

#### Other Changes

- Added `useEarnTokens` hook
This token returns an array of tokens that are supported, enabled, and
where the user is eligible. For example, if pooled-staking is disabled
via remote feature flag, Ethereum will be filtered out of this array.

- Added `Earn%` CTA to supported stablecoins in token list when
stablecoin lending feature flag is enabled.

- Replaced usages of `isStablecoinLendingFeatureEnabled` constant

- Added `MM_POOLED_STAKING_ENABLED`,
`MM_POOLED_STAKING_SERVICE_INTERRUPTION_BANNER_ENABLED`, and
`MM_STABLE_COIN_SERVICE_INTERRUPTION_BANNER_ENABLED` environment
variables

- Removed `MM_STABLECOIN_LENDING_UI_ENABLED` hardcoded env var in
`bitrise.yml`

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Jira ticket: [STAKE-1006: Add earn remote feature flags to
mobile](https://consensyssoftware.atlassian.net/browse/STAKE-1006)

## **Manual testing steps**

1. Locate earn feature flags selectors file -
`app/components/UI/Earn/selectors/featureFlags/index.ts`
2. Manually test by forcing the different selectors to return `true` or
`false`.

### Pooled-Staking UI Changes Based on Flags
`earnPooledStakingEnabled`:
    when disabled:
      - Ethereum "Earn" CTA should be hidden in the token list
- The "Stake" and "Stake more" button should be hidden on the Ethereum
details screen
- If you haven't staked yet, the empty state CTA should be hidden on the
Ethereum details screen

`earnPooledStakingServiceInterruptionBannerEnabled`
    when enabled:
- "We're down for maintenance" banner should be displayed in the "your
earnings section" on the Ethereum details screen.

`earnStablecoinLendingEnabled`
      when enabled:
- The "Earn" CTAs should be displayed for all supported stablecoins in
your token list
- The Earn deposit input screen should be the latest one (with
integrated token selector)
- The "Earn" wallet actions should be displayed in the wallet actions
list

`earnStablecoinLendingServiceInterruptionBannerEnabled`
  - Not in use yet


## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

Cal-L committed:
chore: Update readme and remove e2e smoke from label check (#14900)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

We no longer need to check for `Run E2E Smoke` label in the label gate
since it's enforced by the e2e gate. Removing it from the label gate
reduces the redundancy check

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

ffmcgee725 committed:
Merge branch 'feat/caip-25-refactor' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
refactor: remove solana related implementations migrated from extension (will be done in the next phase)

ffmcgee725 committed:
refactor: remove unused import

sethkfman committed:
chore: add bitrise version bump (#14610)

## **Description**

This PR enables Bitrise to bump the build number version of the app when
sending a version to the store.

It uses the GH Action Update Build Number and will require a new GH
token with permission to read/write actions and content.

This is preferred because it scopes the capability of Bitrise to actions
only instead of write & push access to the repository.

An additional `runway_build_release_candidate` workflow was added to
enable this functionality on Runway builds.

## **Related issues**

Fixes:

## **Manual testing steps**

1. Run any of the modified workflows and see the successful bump version
step
2. Confirm uploads to stores were successful

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

Successful Workflow:
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/4e10f2b6-57e8-464a-bfa1-f509aa4ea560
Successful Runway Workflow:
https://app.bitrise.io/build/78d7fe3a-591d-43db-be70-2a8534529966?tab=log

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: metamaskbot <metamaskbot@users.noreply.github.com>

infiniteflower committed:
fix: Mms 2316 bridge tokens speed cp-7.46.0 (#14904)

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This pull request introduces enhancements to the token selection
components in the bridge feature, focusing on improving the user
experience during loading states and optimizing the handling of top
tokens. Key changes include the addition of a skeleton loader for
partial loading states, updates to the token rendering logic, and
optimizations in the `useTopTokens` hook to limit and prioritize
displayed tokens.

### Enhancements to loading states:
* Introduced a `SkeletonItem` component in `BridgeTokenSelectorBase` to
visually indicate loading states for top tokens. This is used in both
`BridgeSourceTokenSelector` and `BridgeDestTokenSelector` components.
* Updated the `renderTokenItem` and `keyExtractor` logic to handle
`null` items, enabling partial loading with skeleton placeholders.
* Modified the `shouldRenderLoading` logic to differentiate between
overall loading and partial loading, showing skeletons for top tokens
while rendering tokens with balance immediately.

### Optimizations in token prioritization:
* Refactored the `useTopTokens` hook to cap the number of top tokens
displayed at 30 (`MAX_TOP_TOKENS`) and prioritize tokens from feature
flags or swaps API before adding remaining unique tokens. This ensures
better performance and a more curated user experience.

### Minor UI and code improvements:
* Added alignment styles to the `TokenSelectorItem` component for better
layout consistency.
* Simplified string interpolation in `BridgeTokenSelectorBase` to
improve readability.

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->



https://github.com/user-attachments/assets/f243a9ec-0434-45df-b806-e9702a9be1ec



## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

infiniteflower committed:
chore: use Arrow2Left for bridge numpad arrow cp-7.46.0 (#14909)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This pull request includes a small change to the `BridgeView` component.
The change updates the `deleteIcon` property to use the `Arrow2Left`
icon instead of the `ArrowLeft` icon for consistency or design purposes.

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to Bridge
2. Observe keypad is arrow with a line in it

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->


<img width="200" alt="Screenshot 2025-04-25 at 6 49 35 PM"
src="https://github.com/user-attachments/assets/0cf39430-44cd-4039-ba6c-ac53b82934f4"
/>


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

ffmcgee725 committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge remote-tracking branch 'origin/feat/caip-25-refactor' into feat/caip-25-refactor

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
merge origin caip 25 refactor and fix conflicts

ffmcgee725 committed:
feat: successfully using pipeline for multichain rpc handlers

ffmcgee725 committed:
refactor: migration type fixing

ffmcgee725 committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
update migration

ffmcgee725 committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge remote-tracking branch 'origin/feat/caip-25-refactor' into feat/caip-25-refactor

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
fix: fix keyring state issue

ffmcgee725 committed:
test: app/core/RPCMethods/wallet_addEthereumChain.js after merge conflict fix

ffmcgee725 committed:
fix: address SDK connection issue

ffmcgee725 committed:
merge main and fix conflicts

jiexi committed:
yarn dedupe

jiexi committed:
address lint:tsc

ffmcgee725 committed:
merge main fix conflicts

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge remote-tracking branch 'origin/feat/caip-25-refactor' into feat/caip-25-refactor

ffmcgee725 committed:
Merge branch 'main' into feat/caip-25-refactor

ffmcgee725 committed:
merge base branch and fix conflicts

ffmcgee725 committed:
test: fix BackgroundBridge.test.js

ffmcgee725 committed:
test: fix app/util/permissions/index.test.ts

ffmcgee725 committed:
deduplicate

ffmcgee725 committed:
lint

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge remote-tracking branch 'origin/feat/caip-25-refactor' into feat/caip-25-refactor

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

jiexi committed:
Merge branch 'main' into feat/caip-25-refactor

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
Merge branch 'feat/caip-25-refactor' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
test: missing app/selectors/permissions/index.js tests

ffmcgee725 committed:
test: BackgroundBridge increase test cov

abretonc7s committed:
fix: update title assignment in AccountConnect and enhance permission logging

- Changed the title assignment in AccountConnect to directly use channelIdOrHostname for WalletConnect origins.
- Added detailed logging in the updatePermittedChains function to track permission state and changes, improving debugging capabilities.
- Ensured consistent use of hostname for permission checks across WalletConnect sessions and utilities.

ffmcgee725 committed:
Merge branch 'feat/caip-25-refactor' into feat/caip-multichain-migrate-core

abretonc7s committed:
chore: cleanup

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/caip-25-refactor

abretonc7s committed:
refactor: standardize hostname usage in WalletConnect sessions

- Updated the WalletConnect2Session and WC2Manager classes to consistently use the `getHostname` function for URL normalization.
- Enhanced permission handling by integrating `updatePermittedChains` to ensure the current chain is explicitly permitted.
- Improved logging for better debugging and tracking of permission states.

ffmcgee725 committed:
Merge branch 'feat/caip-25-refactor' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
feat: implement multichain local handlers (wip)

ffmcgee725 committed:
merge main fix conflicts

abretonc7s committed:
refactor: Clean up BackgroundBridge and improve test coverage

- Refactored BackgroundBridge.js for better readability and maintainability, including consistent indentation and improved null checks.
- Enhanced BackgroundBridge.test.js with additional mocks and structured tests for middleware configuration and background service management.
- Updated permission differencing utilities tests for clarity and added new test cases for edge scenarios.
- Improved validator utilities tests to cover more cases and ensure robust validation logic.

ffmcgee725 committed:
feat: more local multichain handlers

ffmcgee725 committed:
feat: impl web3shimUsage

ffmcgee725 committed:
debug: log removal

ffmcgee725 committed:
refactor: remove unreferenced code

abretonc7s committed:
test: enhance BackgroundBridge tests and add sendMetadata tests

- Expanded BackgroundBridge.test.js with comprehensive mocks and structured tests for middleware configuration, notification handling, and state management.
- Introduced a new test suite for sendMetadata, validating its implementation and error handling for various parameter scenarios.
- Improved overall test coverage and robustness for the BackgroundBridge functionality.

abretonc7s committed:
wip tests coverage

abretonc7s committed:
wip coverage increase

abretonc7s committed:
wip

abretonc7s committed:
wip

abretonc7s committed:
backup mocking file

abretonc7s committed:
wip coverage

ffmcgee725 committed:
fix: remove unecessary handlers

ffmcgee725 committed:
test: ts lint

ffmcgee725 committed:
test: remove outdated test

ffmcgee725 committed:
test: fix backgroundBridge mocking of unexisting module

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

abretonc7s committed:
feat: address pr todos

ffmcgee725 committed:
refactor: clean up TODOs

ffmcgee725 committed:
lint

ffmcgee725 committed:
docs: remove commented code

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/caip-multichain-migrate-core

abretonc7s committed:
refactor: update multichain API condition and export new authorization utilities

- Modified the condition for initializing the multichain subscription manager to include checks for `isMMSDK` and `isWalletConnect`.
- Exported new utility functions `diffMap`, `getChangedAuthorizations`, and `getRemovedAuthorizations` from the permissions module.

ffmcgee725 committed:
feat: add user facing RPC Methods to caip provider

ffmcgee725 committed:
test: remove BackgroundBridge tests (module to be refactored to TS)

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
merge main, fix conflicts and dedupe

ffmcgee725 committed:
fix: remove user facing RPC methods (test pipeline)

ffmcgee725 committed:
fix: add user facing RPC methods (fixes not working write rpc method calls)

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
docs: remove BACKGROUND_BRIDGE_COVERAGE.md

ffmcgee725 committed:
minor cleanup

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
test: restore BackgroundBridge.test.js

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
refactor: remove unecessary duplicate handlers (watch asset and add ethereum chain)

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

abretonc7s committed:
fix: e2e tests

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

abretonc7s committed:
chore: linting

abretonc7s committed:
fix: enhance validation for typed data in TypedSign component and RPC methods

- Added checks for null or invalid parameters in the TypedSign component to prevent rendering issues.
- Improved error handling in RPC methods for `eth_signTypedData` and `eth_signTypedData_v3/v4` to ensure required fields are present and properly formatted, throwing descriptive errors when validation fails.

abretonc7s committed:
fix: unit tests

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
address pr feedback

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

abretonc7s committed:
chore: linting

ffmcgee725 committed:
refactor: remove tabId from BackgroundBridge.js

abretonc7s committed:
fix: enhance error handling and validation in signature processing

- Added checks to ensure that the message is an array before mapping, returning a simple object if not.
- Improved validation for message types, ensuring proper handling of non-object types.
- Updated error handling in `parseAndSanitizeSignTypedData` and `parseNormalizeAndSanitizeSignTypedData` to catch parsing errors and log warnings while maintaining original behavior of returning an empty object on error.

abretonc7s committed:
fix: improve validation and error handling in signature parsing

- Added checks for missing types and primaryType in `parseAndSanitizeSignTypedData` to enhance validation.
- Updated error handling to allow specific validation errors to propagate while maintaining original behavior for other errors.

ffmcgee725 committed:
Revert "fix: improve validation and error handling in signature parsing"

This reverts commit 9d75e1982dfc07bb2ab0f13b9e47d8756d75967c.

ffmcgee725 committed:
Revert "fix: enhance error handling and validation in signature processing"

This reverts commit 525eed3311243b2ce8123969a4116131e4daac29.

ffmcgee725 committed:
Revert "fix: unit tests"

This reverts commit 9248dd7d91f17ceaed82ad676ba5b3e061f64996.

ffmcgee725 committed:
Revert "fix: enhance validation for typed data in TypedSign component and RPC methods"

This reverts commit c3646db081beeb691b96230da35b492f36b966b0.

ffmcgee725 committed:
revert: signTypedData_v4 only handles stringified JSONs (multichain call not supported)

ffmcgee725 committed:
docs: jsdoc eth_signTypedData_v4 multichain issue

ffmcgee725 committed:
fix: remove iteration through multiple origins (wip)

ffmcgee725 committed:
docs: add todos for removing all iterations through multiple origins

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
lint: tabId not defined

ffmcgee725 committed:
refactor: minor subscription related adjustments (code removal)

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
minor lint fix

ffmcgee725 committed:
revert: comment back permission-system-add-non-permitted-spec.js

ffmcgee725 committed:
refactor: remove more unreferenced code and unecessary exports

ffmcgee725 committed:
feat: add caip25 event subscriptions properly

ffmcgee725 committed:
refactor: minor func name chnage

ffmcgee725 committed:
fix: minor fix to getAuthorizedScopes

ffmcgee725 committed:
test: fix unit test for selector

ffmcgee725 committed:
refactor: pass in origin to getAuthorizedScopes selector to match the correct subject

ffmcgee725 committed:
feat: unsubscribe to CAIP session scope changes on BackgroundBridge disconnect

ffmcgee725 committed:
merge main fix conflicts

ffmcgee725 committed:
refactor: minor tweak to origin

ffmcgee725 committed:
refactor: address pr review comments

jiexi committed:
make handleCaipSessionScopeChanges bound to this context

ffmcgee725 committed:
fix: eth_subscribe fixed

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

ffmcgee725 committed:
lint

ffmcgee725 committed:
Merge branch 'main' into feat/caip-multichain-migrate-core

abretonc7s committed:
feat: add multichain E2E testing guide and support

- Introduced a new document, MULTICHAIN_E2E.md, detailing the setup and execution of multichain end-to-end tests for MetaMask Mobile.
- Added the MultichainTestDApp page object to facilitate interactions with the multichain test dapp.
- Updated fixture-helper.js to support multichain configurations and introduced DEFAULT_MULTICHAIN_TEST_DAPP_FIXTURE_OPTIONS.
- Included new multichain test examples in e2e/specs/multichain/multichain-api-example.spec.js.
- Updated package.json and yarn.lock to include the @metamask/test-dapp-multichain dependency.

abretonc7s committed:
feat: enhance multichain test dapp support and update E2E tests

- Added URL handling for multichain test dapp resources in create-static-server.js.
- Updated fixture-helper.js to replace Anvil with Ganache for local node options.
- Refactored multichain API tests to simplify the connection process and verify dapp loading without creating a session.
- Improved test assertions for better validation of multichain dapp functionality.

abretonc7s committed:
refactor: update MULTICHAIN_E2E.md for clarity and focus

- Removed redundant information about the multichain test dapp setup.
- Enhanced the section on testing goals to specify EVM network focus and core test modules.
- Updated the running tests section to include specific commands for debugging.
- Added troubleshooting tips for multichain test dapp loading issues.

abretonc7s committed:
feat: enhance MultichainTestDApp functionality and add new selectors

- Refactored MultichainTestDApp class to improve WebView interactions and added new methods for connecting, creating sessions, and revoking sessions.
- Introduced MultichainTestDappSelectorsWebIDs for better management of WebView element IDs.
- Updated multichain API tests to verify loading, connection, and session creation with multiple chains.
- Improved assertions for better validation of the multichain dapp functionality.

abretonc7s committed:
feat: migrate MultichainTestDApp to TypeScript and enhance test coverage

- Converted MultichainTestDApp from JavaScript to TypeScript for improved type safety and maintainability.
- Added new selectors in TypeScript for better management of WebView element IDs.
- Updated Jest configuration to include TypeScript test files.
- Removed outdated JavaScript files and ensured all references are updated accordingly.
- Introduced new E2E tests for multichain functionality, verifying loading, connection, and session management.

abretonc7s committed:
feat: enhance MultichainTestDApp with new connection methods and improved flow

- Added new methods for connecting to the dapp using JavaScript injection and auto-connect button.
- Refactored existing connection logic to streamline the multichain connection process.
- Reduced delay times for various operations to improve performance.
- Updated E2E tests to utilize the new connection methods and verify custom chain selections.
- Enhanced error handling and removed console logs for cleaner output.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
chore: update multichain dapp

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
feat: Enhance multichain session management and add comprehensive E2E tests

- Introduced new methods in `MultichainTestDApp` for session management, including `revokeSessionWithData`, `selectNetworks`, and `createSessionWithNetworks`.
- Improved error handling with detailed console logs for session operations.
- Added new E2E test files for `wallet_createSession`, `wallet_getSession`, `wallet_revokeSession`, and `wallet_invokeMethod` APIs, covering various scenarios including session creation, retrieval, and revocation.
- Utilized `MultichainUtilities` for consistent handling of chain IDs and session assertions across tests.
- Enhanced test coverage for multichain interactions, ensuring robust validation of session states and method invocations.

abretonc7s committed:
feat: Add daily priorities and multichain E2E test progress documentation

- Created `DAILY_PRIORITIES_MULTICHAIN_E2E.md` to outline daily testing goals, completed tasks, and current blockers for multichain E2E testing.
- Introduced `MULTICHAIN_E2E_PROGRESS.md` to provide a comprehensive overview of the progress on multichain E2E tests, including completed work, current tasks, and known issues.
- Added `MULTICHAIN_E2E_TEST_RESULTS.md` to summarize test results, highlighting overall status, detailed test outcomes, and critical issues affecting test execution.
- Enhanced documentation for better tracking of multichain testing efforts and to facilitate collaboration among team members.

abretonc7s committed:
feat: Complete multichain E2E tests and documentation

- Added comprehensive E2E tests for all 6 core multichain APIs: `wallet_createSession`, `wallet_getSession`, `wallet_revokeSession`, `wallet_invokeMethod`, `wallet_notify`, and `wallet_sessionChanged`.
- Achieved 100% API coverage with detailed test scenarios and improved error handling.
- Introduced `MULTICHAIN_E2E_FINAL_SUMMARY.md` and `MULTICHAIN_E2E_NEXT_STEPS.md` to document project completion status, implementation details, and future testing priorities.
- Enhanced `MULTICHAIN_E2E_PROGRESS.md` and `MULTICHAIN_E2E_TEST_RESULTS.md` to reflect the final test outcomes and overall project success.
- Implemented native Detox selectors for reliable WebView interactions, addressing previous UI issues with method selection modals.

abretonc7s committed:
fix: Clean up whitespace in test files and selectors

- Removed unnecessary blank lines and trailing spaces in various test files and selectors to improve code readability and maintainability.
- Ensured consistent formatting across the codebase, particularly in the `MultichainTestDApp` and related test specifications.

abretonc7s committed:
feat: Update multichain test configuration and documentation

- Added multichain test dapp URL configuration options to `.e2e.env.example` for improved flexibility in testing environments.
- Created `MULTICHAIN_DAPP_CONFIG.md` to provide detailed instructions on configuring the multichain test dapp URL for E2E testing.
- Removed outdated files: `DAILY_PRIORITIES_MULTICHAIN_E2E.md`, `MULTICHAIN_E2E_FINAL_SUMMARY.md`, `MULTICHAIN_E2E_NEXT_STEPS.md`, `MULTICHAIN_E2E_PROGRESS.md`, and `MULTICHAIN_E2E_TEST_RESULTS.md` to streamline documentation.
- Enhanced `MultichainTestDApp` to dynamically select the test dapp URL based on environment variables, improving usability for developers.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
chore: Remove outdated MULTICHAIN_DAPP_CONFIG.md and enhance MULTICHAIN_E2E_RESULTS.md

- Deleted the obsolete MULTICHAIN_DAPP_CONFIG.md file to streamline documentation.
- Updated MULTICHAIN_E2E_RESULTS.md to include environment loading instructions and detailed multichain test dapp configuration options for improved clarity and usability.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
feat: add SmokeMultichainApi tag and update related tests

- Introduced a new tag `smokeMultichainApi` in `tags.js`.
- Updated multiple test files to use the new `SmokeMultichainApi` tag instead of the deprecated `SmokeNetworkExpansion` for better clarity and organization in multichain API tests.

abretonc7s committed:
refactor: remove console logs from multichain wallet tests for cleaner output

- Eliminated unnecessary console log statements from `wallet-createSession.spec.ts`, `wallet-invokeMethod.spec.ts`, `wallet-notify.spec.ts`, and `wallet-sessionChanged.spec.ts` to enhance test output clarity.
- Updated test descriptions for better understanding of test objectives.
- Ensured that the tests maintain their functionality while improving readability and reducing noise in the logs.

ffmcgee725 commented on e2e/fixtures/fixture-helper.js:
```diff
@@ -289,15 +301,10 @@ export async function withFixtures(options, testSuite) {
-            'test-dapp',
-            'dist',
-          );
+          // Select dapp directory based on multichainDapp flag
```
we can remove comment

ffmcgee725 commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,838 @@
+      return { success: false };
+    } catch (error) {
+      console.error('❌ Error getting session data:', error);
+      return { success: false };
```
same here

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
chore: update dapp dependencies

abretonc7s committed:
fix: pr comments

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
chore: update multichain dapp

abretonc7s committed:
chore: cleanup

abretonc7s committed:
chore: cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
refactor: update wallet session change test to use Base network instead of Arbitrum

- Removed initial sessionChanged history count check.
- Updated the modified networks to include Base network instead of Arbitrum.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
feat: enhance multichain test dapp and wallet notification tests

- Added a default multichain test dapp URL for improved accessibility.
- Updated the logic in `getMultichainTestDappUrl` to default to the new URL if local development is not explicitly set to false.
- Expanded the `wallet_notify` E2E tests to include detailed steps for checking notification states and subscription results, ensuring robust validation of real-time event handling.
- Refined the `wallet_sessionChanged` tests to focus on verifying session modifications when networks are added, enhancing clarity and accuracy in event assertions.

abretonc7s committed:
feat: enhance multichain wallet invoke method tests

- Improved result verification for `eth_chainId`, `eth_getBalance`, and `eth_gasPrice` methods, ensuring accurate logging of RPC results.
- Added checks for visibility of invoke containers and result elements, including handling for truncated results.
- Enhanced logging for better debugging and clarity during test execution.
- Updated delays to ensure sufficient time for method execution and result retrieval.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
feat: latest multichain dapp

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

jiexi commented on .e2e.env.example:
```diff
@@ -5,3 +5,11 @@ export MM_TEST_ACCOUNT_PRIVATE_KEY=''
+# Option 2: Use a custom remote URL
+# export MULTICHAIN_DAPP_URL="https://your-multichain-dapp.example.com/"
+# Option 3: Use local development server
+# export USE_LOCAL_DAPP=true
```
IMO this should be the default

abretonc7s replied:
this is the default actually, I havent updated the example. Good catch

jiexi commented on e2e/jest.e2e.config.js:
```diff
@@ -8,6 +8,10 @@ module.exports = {
+    '<rootDir>/e2e/specs/*.spec.ts',
+    '<rootDir>/e2e/specs/*/*.spec.ts',
+    '<rootDir>/e2e/specs/*/*/*.spec.ts',
+    '<rootDir>/e2e/specs/*/*/*/*.spec.ts',
```
mmm... we could probably use `**` glob here, but whatever

jiexi commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,684 @@
+  /**
+   * Connect to the dapp by entering extension ID and clicking connect
+   */
+  async connect(_extensionId = 'window.postMessage'): Promise<boolean> {
```
would clicking the `auto connect with postMessage` button not be more reliable than this?

jiexi commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,684 @@
+    await this.tapButton(createSessionBtn);
+
+    // Wait for session creation
+    await TestHelpers.delay(1500);
```
is there a way to wait for the confirmation to show up instead?

abretonc7s replied:
Yeah waiting for elements in detox webview is super flaky. I've found fixed delays are way more reliable than trying to wait for specific elements to appear/disappear in the webview context. The webview bridge timing is unpredictable and element queries often fail even when the element is there.

jiexi commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,684 @@
+    await this.tapButton(getSessionBtn);
+
+    // Wait for processing
+    await TestHelpers.delay(1000);
```
guessing it's not really possible to wait for an element to appear on the page

abretonc7s replied:
It is very limited with detox on the webview. We can do it on normal react native ui but somehow was unreliable in the webview

jiexi commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,684 @@
+    const createSessionButton = webview.element(by.web.id('create-session-btn'));
+
+    await createSessionButton.scrollToView();
+    await createSessionButton.runScript('(el) => { el.click(); }');
```
is there a difference?

abretonc7s replied:
the direct webview approach seems a bit more consistent on web element but again it depends on the situation and the component since it is generally fragile anything with the webview. For example using CSS :checked selector is the only reliable way I found for checkbox states, and you can do fallback patterns like checkbox.tap() then runScript('el.click()') if needed. The tapButton method is simpler but less flexible when you need JS-level control

abretonc7s committed:
refactor: simplify testMatch patterns in jest.e2e.config.js

abretonc7s committed:
cleanup

abretonc7s committed:
chore: update .e2e.env.example to configure local and remote dapp usage

abretonc7s committed:
pr comments

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
chore: update .depcheckrc.yml to ignore '@metamask/test-dapp-multichain'

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
linter

ffmcgee725 commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,684 @@
+  async selectNetwork(chainId: string): Promise<boolean> {
+    const webview = this.getWebView();
+    // Escape colons in chain ID to match the dapp's escapeHtmlId function
+    const escapedChainId = `eip155:${chainId}`.replace(/:/g, '-');
```
➕

abretonc7s replied:
done!

ffmcgee725 commented on package.json:
```diff
@@ -125,7 +125,7 @@
     "crypto-js": "4.2.0",
     "axios": "^1.8.2",
-
+    
```
remove empty space

abretonc7s commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,621 @@
+      };
+    }
+
+    return { success: true };
```
not necessarily it is optional

abretonc7s committed:
pr comments

abretonc7s committed:
pr comments

abretonc7s committed:
feat: new ui pattern

abretonc7s committed:
pr comments

abretonc7s reviewed (commented):
pr comments

ffmcgee725 commented on e2e/specs/multichain/wallet-invokeMethod.spec.ts:
```diff
@@ -0,0 +1,497 @@
+                            // Could not read result text (Detox limitation)
+                        }
+                    }
+
+
```
unnecessary empty spaces

ffmcgee725 commented on e2e/specs/multichain/wallet-invokeMethod.spec.ts:
```diff
@@ -0,0 +1,497 @@
+                        const resultElement = webview.element(by.web.id(resultElementId));
+                        await resultElement.scrollToView();
+
+                        try {
+                            const resultText = await resultElement.runScript((el) => el.textContent || '');
+                            console.log(`eth_gasPrice result: ${resultText}`);
+                            if (method === 'eth_gasPrice' && resultText.includes('"0x')) {
+                                console.log('✅ eth_gasPrice returned a valid hex gas price');
+                            }
+                        } catch (readError) {
+                            // Could not read result text (Detox limitation)
+                        }
```
does this need to be inside a nested `try..catch` ?

abretonc7s committed:
formatting

ffmcgee725 commented on e2e/specs/multichain/wallet-invokeMethod.spec.ts:
```diff
@@ -0,0 +1,497 @@
+                            // Could not read result text (Detox limitation)
+                        }
+                    }
+
+
```
unnecessary empty spaces

ffmcgee725 commented on e2e/specs/multichain/wallet-invokeMethod.spec.ts:
```diff
@@ -0,0 +1,497 @@
+                            const directButton = webview.element(by.web.id(directButtonId));
+                            await directButton.tap();
+
+                            try {
+                                // Both eth_sendTransaction and personal_sign use Cancel button
+                                const cancelButton = element(by.text('Cancel'));
+                                await waitFor(cancelButton).toBeVisible().withTimeout(10000);
+                                await cancelButton.tap();
+                                console.log(`✅ ${method} confirmation screen test passed`);
+
+                            } catch (confirmError) {
+                                throw new Error(`${method} should have triggered confirmation screen but didn't`);
+                            }
+
+                        } catch (methodError) {
+                            throw new Error(`${method} method invocation failed - this is a test failure`);
+                        }
```
Do we need to nest this in a `try..catch` ?

ffmcgee725 commented on e2e/specs/multichain/wallet-invokeMethod.spec.ts:
```diff
@@ -0,0 +1,497 @@
+
+                        await TestHelpers.delay(3000);
+
+                        try {
+                            const detailsId = `method-result-details-${escapedScopeForButton}-${method}-0`;
+                            const detailsElement = webview.element(by.web.id(detailsId));
+                            await detailsElement.scrollToView();
+
+                            try {
+                                const resultElementId = `invoke-method-${escapedScopeForButton}-${method}-result-0`;
+                                const resultElement = webview.element(by.web.id(resultElementId));
+                                const resultText = await resultElement.runScript((el) => el.textContent || '');
+                                console.log(`${method} result: ${resultText}`);
+                            } catch (e) {
+                                // Could not read result text
+                            }
```
same

abretonc7s committed:
pr comments

abretonc7s committed:
cleanup

jiexi commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,657 @@
+    }
+
+    for (const chainId of expectedChainIds) {
+      const fullChainId = chainId.startsWith('eip155:') ? chainId : `eip155:${chainId}`;
```
thoughts on typing `expectedChainIds` to CaipChainId?

abretonc7s replied:
This way can handle both

jiexi commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,657 @@
+    const webview = this.getWebView();
+    const allNetworks = this.getAllSupportedNetworks();
+
+          for (const chainId of allNetworks) {
```
lint

jiexi commented on e2e/specs/multichain/wallet-sessionChanged.spec.ts:
```diff
@@ -0,0 +1,95 @@
+        jest.setTimeout(150000);
+    });
+
+    it('should receive a wallet_sessionChanged event when a new network is added to the session', async () => {
```
this title isn't entirely correct since you've actually made a new session rather than edit the existing via the permission editing UI

jiexi commented on e2e/specs/multichain/wallet-revokeSession.spec.ts:
```diff
@@ -0,0 +1,334 @@
+        );
+    });
+
+    it('should handle multiple revoke calls gracefully', async () => {
```
seems similar to `should handle revoke session when no session exists` except that it does it two more times

abretonc7s replied:
duplicated

abretonc7s committed:
duplicated test

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
pr comments

adonesky1 reviewed (dismissed):
Great work @abretonc7s 🙇

cortisiko commented:
Hi, this not merge-ready. There are e2e tests added that are not running as part of any pipeline. I also see that no QA engineer reviewed these tests. Can we also make sure to include QA in e2e test PRs. Especially if it is the first time we are adding new tests to the repo.  I am going to remove them from the merge queue. 

See my message here explaining how the e2e is running on CI: https://consensys.slack.com/archives/C02U025CVU4/p1744833958446979

abretonc7s committed:
Merge branch 'main' into feat/multichain-e2e

adonesky1 committed:
SmokeMultichainApi->SmokeNetworkExpansion

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
update multichain dapp to latest

abretonc7s committed:
Enhance MultichainTestDApp methods to support result index parameter for session data retrieval. Updated revokeSessionWithData and getSessionData methods to accept an index for results, improving flexibility in handling multiple session results. Adjusted clickFirstResultSummary to handle dynamic result indices and added error handling for non-existent results. Updated related test cases to reflect changes in network selection.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

chrisleewilcox commented on e2e/pages/Browser/MultichainTestDApp.ts:
```diff
@@ -0,0 +1,684 @@
+  /**
+   * Click the create session button
+   */
+  async clickCreateSessionButton(): Promise<boolean> {
```
Established pattern for mobile e2e is to start PO actions with tap instead of click.

chrisleewilcox commented on e2e/specs/multichain/wallet-createSession.spec.ts:
```diff
@@ -0,0 +1,308 @@
+                restartDevice: true,
+            },
+            async () => {
+                await TestHelpers.reverseServerPort();
+                await loginToApp();
+                await TabBarComponent.tapBrowser();
+                await Assertions.checkIfVisible(Browser.browserScreenID);
+                await MultichainTestDApp.navigateToMultichainTestDApp('?autoMode=true');
+
+                await Assertions.checkIfVisible(
+                    Promise.resolve(element(by.id(BrowserViewSelectorsIDs.BROWSER_WEBVIEW_ID))),
+                );
```
Section is repeated in all tests in this file.  Maybe we can have a convenience/helper function for this.

chrisleewilcox commented on e2e/specs/multichain/wallet-createSession.spec.ts:
```diff
@@ -0,0 +1,308 @@
+                    MultichainUtilities.NETWORK_COMBINATIONS.SINGLE_ETHEREUM
+                );
+
+                const assertions = MultichainUtilities.generateSessionAssertions(
+                    sessionResult,
+                    MultichainUtilities.NETWORK_COMBINATIONS.SINGLE_ETHEREUM
+                );
+
+                MultichainUtilities.logSessionDetails(sessionResult, 'Single Ethereum Test');
+
+                if (!assertions.structureValid) {
+                    throw new Error('Invalid session structure');
+                }
+
+                if (!assertions.success) {
+                    throw new Error('Session creation failed');
+                }
+
+                if (!assertions.chainsValid) {
+                    throw new Error(`Chain validation failed. Missing chains: ${assertions.missingChains.join(', ')}`);
+                }
+
+                if (assertions.chainCount !== 1) {
+                    throw new Error(`Expected 1 chain, but found ${assertions.chainCount}`);
+                }
```
Consider using our SoftAssert class that will allow multiple assertions without ending the test on the first failure.

abretonc7s committed:
feat: enhance MultichainTestDApp with new setup method and refactor button interactions

- Added `setupAndNavigateToTestDapp` method for streamlined test setup, including login and navigation.
- Refactored button interaction methods to use `tap` instead of `click`, improving consistency.
- Updated session handling methods to throw errors on failure, enhancing error management.
- Adjusted related test cases to utilize the new setup method and ensure proper session data retrieval.

abretonc7s commented on e2e/specs/multichain/wallet-createSession.spec.ts:
```diff
@@ -0,0 +1,308 @@
+    });
+
+    it('should create a session with Ethereum mainnet scope', async () => {
+        await withFixtures(
+            {
+                ...DEFAULT_MULTICHAIN_TEST_DAPP_FIXTURE_OPTIONS,
+                fixture: new FixtureBuilder().withPopularNetworks().build(),
+                restartDevice: true,
+            },
```
https://github.com/MetaMask/metamask-mobile/pull/15483/commits/7e07ac00c974aefeff3af5376aeee5a6e5687972

abretonc7s commented on e2e/specs/multichain/wallet-createSession.spec.ts:
```diff
@@ -0,0 +1,308 @@
+
+                const sessionResult = await MultichainTestDApp.createSessionWithNetworks(
+                    MultichainUtilities.NETWORK_COMBINATIONS.SINGLE_ETHEREUM
+                );
```
https://github.com/MetaMask/metamask-mobile/pull/15483/commits/7e07ac00c974aefeff3af5376aeee5a6e5687972

chrisleewilcox commented:
Mobile e2e guidelines: https://github.com/MetaMask/contributor-docs/blob/main/docs/testing/e2e/mobile-e2e-guidelines.md

abretonc7s committed:
fix: improve element interaction handling in MultichainTestDApp

- Updated tap and scroll methods to check for element existence before interaction, preventing errors when elements are not present.
- Enhanced comments for clarity on the behavior of the tap action when elements are missing.

abretonc7s committed:
fix: streamline error handling in tapFirstResultSummary method

- Refactored the tapFirstResultSummary method to simplify error handling when the target element does not exist, allowing for smoother execution without unnecessary checks.
- Enhanced comments to clarify the expected behavior when elements are missing.

abretonc7s committed:
fix: restore iOS project file from main branch

abretonc7s committed:
refactor: update MultichainTestDApp and selectors for improved functionality

- Refactored MultichainTestDApp to utilize new selector IDs and added methods for session management and event handling.
- Enhanced selectors to include additional IDs and timeout constants for better test reliability.
- Updated test specifications to streamline method invocation and notification handling, ensuring consistent behavior across different scenarios.
- Improved error handling and logging for better debugging during test execution.

abretonc7s committed:
refactor: enhance MultichainTestDApp and test specifications for improved reliability

- Updated MultichainTestDApp to interact with WebView elements more effectively, removing unnecessary visibility checks.
- Improved test assertions for session creation and retrieval, ensuring accurate failure conditions are validated.
- Enhanced error handling in method invocation tests, providing clearer feedback on failures and expected results.
- Adjusted comments for clarity and to reflect the updated logic in test cases.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/multichain-e2e

abretonc7s committed:
remove unhelpful tests

abretonc7s committed:
Merge branch 'main' into feat/multichain-e2e

abretonc7s committed:
Merge branch 'main' into feat/multichain-e2e

chrisleewilcox committed:
fix: adjust webview retrieval for iOS platform in MultichainTestDApp

chrisleewilcox committed:
refactor: streamline webview retrieval in MultichainTestDApp using Matchers

chrisleewilcox committed:
feat: add multichain API smoke tests workflow for iOS (#16434)

- Introduced new smoke test workflows for the multichain API in the
Bitrise configuration.
- Updated tags to include 'SmokeMultiChainAPI' for better test
categorization.
- Refactored existing test cases to utilize the new multichain API tag
for session management tests.

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.
